rse-py-scripting.Rmd

```diff
@@ -0,0 +1,143 @@
+that works like any other Unix command-line tool.
+
+In this chapter we take the code we developed earlier (TODO: Link)
+and turn it into a Python script (i.e. a command line program written in Python).
+
+
+## A template script {#rse-py-scripting-template}
+
+In general, the first thing that gets added to any Python script is the following:
+
+```python
+if __name__ == '__main__':
+    main()
+```
+
+The reason we need these two lines of code
+is that running a Python script in bash is very similar to importing that file in Python. 
+The biggest difference is that we don’t expect anything to happen when we import a file, 
+whereas when running a script we expect to see some output
+(e.g. an output file, figure and/or some text printed to the screen).
+
+The `__name__` variable exists to handle these two situations.
+When you import a Python file `__name__` is set to the name of that file
+(e.g. when importing script.py, `__name__` is `script`),
+but when running a script in bash `__name__` is always set to `__main__`.
+The convention is to call the function that produces the output `main()`,
+but you can call it whatever you like.
+
+The next thing you'll need is a library to parse the command line for input arguments.
+The most widely used option is 
+[argparse](https://docs.python.org/3/library/argparse.html).
+
+Putting those together,
+here's a basic python script for us to start with:
+
+```shell
+$ cat script_template.py
+```
+
+```python
+import argparse
+
+#
+# All your functions (that will be called by main()) go here.
+#
+
+def main(inargs):
+    """Run the program."""
+
+    print('Input file: ', inargs.infile)
+    print('Output file: ', inargs.outfile)
+
+
+if __name__ == '__main__':
+
+    description='Print the input arguments to the screen.'
+    parser = argparse.ArgumentParser(description=description)
+    
+    parser.add_argument("infile", type=str, help="Input file name")
+    parser.add_argument("outfile", type=str, help="Output file name")
+
+    args = parser.parse_args()            
+    main(args)
+```
+
+By running `script_template.py` at the command line
+we'll see that `argparse` handles all the input arguments:
+
+```shell
+$ python script_template.py in.csv out.png
+```
+
+```text
+Input file:  in.csv
+Output file:  out.png
+```
+
+It also generates help information for the user:
+
+```shell
+$ python script_template.py -h
+```
+
+```text
+usage: script_template.py [-h] infile outfile
+
+Print the input arguments to the screen.
+
+positional arguments:
+  infile      Input file name
+  outfile     Output file name
+
+optional arguments:
+  -h, --help  show this help message and exit
+```
+
+and issues errors when users give the program invalid arguments:
+
+```shell
+$ python script_template.py in.csv
+```
+
+```text
+usage: script_template.py [-h] infile outfile
+script_template.py: error: the following arguments are required: outfile
+```
+
+## Adding our code {#rse-py-scripting-code}
```

Comment:
Should we cover optional and default arguments before this or it that outside the scope of this book?

Reply:
I think in the process of "adding our code" we should introduce optional and default arguments.